presto-main/src/main/java/com/facebook/presto/operator/repartition/ArrayBlockEncodingBuffer.java

```diff
@@ -259,17 +259,13 @@ private void populateNestedPositions(ColumnarArray columnarArray)
         int[] positions = getPositions();
 
         for (int i = 0; i < positionCount; i++) {
-            int position = positions[i];
-            int beginOffset = columnarArray.getOffset(position);
-            int endOffset = columnarArray.getOffset(position + 1);
-            int length = endOffset - beginOffset;
+            offsets[i + 1] = offsets[i] + columnarArray.getLength(positions[i]);
+        }
 
-            offsets[i + 1] = offsets[i] + length;
+        valuesBuffers.setPositions(ensureCapacity(valuesBuffers.getPositions(), offsets[positionCount], SMALL, NONE, bufferAllocator));
```

Comment:
@yingsu00 perhaps, a cleaner design would be `valuesBuffers.ensureCapacity(offsets[positionCount])`

Reply:
@mbasmanova Hi Masha, thanks for the suggestion. I actually thought of this approach before, but was hesitant to use it because the buffers inside the BlockEncodingBuffer is better to be hidden within the class. By using `valuesBuffers.ensureCapacity(offsets[positionCount])`, the caller(user) has to know what int[] or byte[] buffers are inside each of the BlockEncodingBuffer and we may have to identify which buffer to ensureCapacity on, e.g.  `valuesBuffers.ensurePositionsCapacity() `or `valuesBuffers.ensureOffsetsCapacity()`, or `valuesBuffers.ensureCapacity(bufferType, capacity)`. On the other hand, I think adding a `setPositions` method is fair since we already had `getPositions`. What do you think? Do you strongly recommend to use ` valuesBuffers.ensureCapacity(offsets[positionCount])`?

Reply:
@yingsu00 It is generally better to avoid having setters in classes other than POJOs, hence, my preference is to create narrow mutators.

Reply:
@mbasmanova ok, so I will add the `valuesBuffers.ensureOffsetsCapacity(capacity)` method to `AbstractBlockEncodingBuffer`. Does that sounds good?

Reply:
@yingsu00 Shouldn't it be `ensurePositionsCapacity` to match `appendPositionRange`? The contract would be that the caller should call `ensurePositionsCapacity` for the total number of positions that will be appended using `appendPositionRange`.